File: frontend/src/services/textService.ts
/**
 * Typing Text Service
 */
import { apiClient } from './api';
import type { TypingText } from '@/types';

interface TypingTextResponse {
  id: string;
  content: string;
  difficulty: string;
  length: string;
  word_count: number;
  category: string | null;
  author: string | null;
}

interface TextListResponse {
  texts: TypingTextResponse[];
  total: number;
  page: number;
  pages: number;
}

function mapTextResponse(response: TypingTextResponse): TypingText {
  return {
    id: response.id,
    content: response.content,
    difficulty: response.difficulty as 'easy' | 'medium' | 'hard',
    length: response.length as 'short' | 'medium' | 'long',
    wordCount: response.word_count,
    category: response.category || undefined,
  };
}

export const textService = {
  async getRandomText(options?: { difficulty?: string; length?: string }): Promise<TypingText> {
    const params = new URLSearchParams();
    if (options?.difficulty) params.append('difficulty', options.difficulty);
    if (options?.length) params.append('length', options.length);

    const queryString = params.toString();
    const endpoint = queryString ? `/texts/random?${queryString}` : '/texts/random';

    const response = await apiClient.get<TypingTextResponse>(endpoint);
    return mapTextResponse(response);
  },

  async listTexts(options?: {
    page?: number;
    limit?: number;
    difficulty?: string;
    length?: string;
  }): Promise<{ texts: TypingText[]; total: number; page: number; pages: number }> {
    const params = new URLSearchParams();
    if (options?.page) params.append('page', options.page.toString());
    if (options?.limit) params.append('limit', options.limit.toString());
    if (options?.difficulty) params.append('difficulty', options.difficulty);
    if (options?.length) params.append('length', options.length);

    const queryString = params.toString();
    const endpoint = queryString ? `/texts?${queryString}` : '/texts';

    const response = await apiClient.get<TextListResponse>(endpoint);
    return {
      texts: response.texts.map(mapTextResponse),
      total: response.total,
      page: response.page,
      pages: response.pages,
    };
  },

  async getText(textId: string): Promise<TypingText> {
    const response = await apiClient.get<TypingTextResponse>(`/texts/${textId}`);
    return mapTextResponse(response);
  },
};

File: backend/app/application/services/text_service.py
"""
Typing text service.

Handles text retrieval and management.
"""
from typing import Optional
from uuid import UUID

from app.domain.entities.typing_text import Difficulty, TextLength, TypingText
from app.domain.exceptions import EntityNotFoundError
from app.domain.repositories.text_repository import TextRepository


class TextService:
    """Service handling typing text operations."""

    def __init__(self, text_repository: TextRepository):
        self._text_repo = text_repository

    async def get_random_text(
        self,
        difficulty: Optional[str] = None,
        length: Optional[str] = None,
    ) -> TypingText:
        """
        Get a random text with optional filters.

        Args:
            difficulty: Optional difficulty filter.
            length: Optional length filter.

        Returns:
            Random typing text.

        Raises:
            EntityNotFoundError: If no matching text found.
        """
        diff = Difficulty(difficulty) if difficulty else None
        len_ = TextLength(length) if length else None

        text = await self._text_repo.get_random(difficulty=diff, length=len_)

        if not text:
            raise EntityNotFoundError("TypingText", f"{difficulty}/{length}")

        return text

    async def get_text_by_id(self, text_id: UUID) -> TypingText:
        """
        Get a specific text by ID.

        Args:
            text_id: Text's unique identifier.

        Returns:
            Typing text.

        Raises:
            EntityNotFoundError: If text not found.
        """
        text = await self._text_repo.get_by_id(text_id)
        if not text:
            raise EntityNotFoundError("TypingText", str(text_id))
        return text

    async def list_texts(
        self,
        difficulty: Optional[str] = None,
        length: Optional[str] = None,
        page: int = 1,
        limit: int = 20,
    ) -> dict:
        """
        List texts with pagination and filters.

        Args:
            difficulty: Optional difficulty filter.
            length: Optional length filter.
            page: Page number.
            limit: Results per page.

        Returns:
            Dictionary with texts and pagination info.
        """
        diff = Difficulty(difficulty) if difficulty else None
        len_ = TextLength(length) if length else None
        offset = (page - 1) * limit

        texts, total = await self._text_repo.get_all(
            difficulty=diff,
            length=len_,
            limit=limit,
            offset=offset,
        )

        return {
            "texts": [
                {
                    "id": str(t.id),
                    "content": t.content,
                    "difficulty": t.difficulty.value,
                    "length": t.length.value,
                    "word_count": t.word_count,
                    "category": t.category,
                    "author": t.author,
                }
                for t in texts
            ],
            "total": total,
            "page": page,
            "pages": (total + limit - 1) // limit,
        }
